datafusion/src/optimizer/tokomak/datatype.rs

```diff
@@ -0,0 +1,166 @@
+use crate::error::DataFusionError;
+use arrow::datatypes::{DataType, IntervalUnit, TimeUnit};
+use egg::*;
+use std::convert::TryFrom;
+use std::fmt::Display;
+use std::str::FromStr;
+
+define_language! {
+    #[derive(Copy)]
+    pub enum TokomakDataType {
+        "date32" = Date32,
+        "date64" = Date64,
+        "bool" = Boolean,
+        "int8" = Int8,
+        "int16" =Int16,
+        "int32" =Int32,
+        "int64" =Int64,
+        "uint8" =UInt8,
+        "uint16" =UInt16,
+        "uint32" =UInt32,
+        "uint64" =UInt64,
+        "float16" =Float16,
+        "float32" =Float32,
+        "float64" =Float64,
+        "utf8"=Utf8,
+        "largeutf8"=LargeUtf8,
+        "time(s)"=TimestampSecond,
+        "time(ms)"=TimestampMillisecond,
+        "time(us)"=TimestampMicrosecond,
+        "time(ns)"=TimestampNanosecond,
+        "interval(yearmonth)"=IntervalYearMonth,
+        "interval(daytime)"=IntervalDayTime,
+    }
+}
+
+impl Display for TokomakDataType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!("{:?}", self))
+    }
+}
+
+impl Into<DataType> for &TokomakDataType {
+    fn into(self) -> DataType {
+        let v = *self;
+        v.into()
+    }
+}
+
+impl Into<DataType> for TokomakDataType {
+    fn into(self) -> DataType {
+        match self {
+            TokomakDataType::Date32 => DataType::Date32,
+            TokomakDataType::Date64 => DataType::Date64,
+            TokomakDataType::Boolean => DataType::Boolean,
+            TokomakDataType::Int8 => DataType::Int8,
+            TokomakDataType::Int16 => DataType::Int16,
+            TokomakDataType::Int32 => DataType::Int32,
+            TokomakDataType::Int64 => DataType::Int64,
+            TokomakDataType::UInt8 => DataType::UInt8,
+            TokomakDataType::UInt16 => DataType::UInt16,
+            TokomakDataType::UInt32 => DataType::UInt32,
+            TokomakDataType::UInt64 => DataType::UInt64,
+            TokomakDataType::Float16 => DataType::Float16,
+            TokomakDataType::Float32 => DataType::Float32,
+            TokomakDataType::Float64 => DataType::Float64,
+            TokomakDataType::Utf8 => DataType::Utf8,
+            TokomakDataType::LargeUtf8 => DataType::LargeUtf8,
+            TokomakDataType::TimestampSecond => {
+                DataType::Timestamp(TimeUnit::Second, None)
+            }
+            TokomakDataType::TimestampMillisecond => {
+                DataType::Timestamp(TimeUnit::Millisecond, None)
+            }
+            TokomakDataType::TimestampMicrosecond => {
+                DataType::Timestamp(TimeUnit::Microsecond, None)
+            }
+            TokomakDataType::TimestampNanosecond => {
+                DataType::Timestamp(TimeUnit::Nanosecond, None)
```

Comment:
(no action needed, just a note) IMO we should solve this Timestamp with timezone ASAP: the arrow format supports timezone, and here we are already assuming that we not use it. However, this is tricky because we are just making it harder to support timezones.

Note that the "big" problem with timezones is that the `DataType::Timestamp()` is virtually uncountable, which usually makes typing systems a bit difficult to build on top of.